File: webapp/backend/app/schemas/anomaly.py
from __future__ import annotations

from typing import Dict, List, Optional

from pydantic import BaseModel, Field

from .common import BaseResponse


class Evidence(BaseModel):
    type: str
    strength: float
    detail: str
    ref: Optional[str] = None


class AnomalyThresholds(BaseModel):
    point: float
    pattern: float
    trend: float


class AnomalyDetectRequest(BaseModel):
    product_id: str
    aps: str
    date_range: Optional[List[str]] = None
    thresholds: Optional[AnomalyThresholds] = None
    threshold: Optional[float] = None
    include_explanations: bool = True


class AnomalySeriesItem(BaseModel):
    date: str
    demand: float
    expected: float
    residual: float
    anomaly_score: float
    is_anomaly: bool
    anomaly_family: Optional[str] = None
    anomaly_type: Optional[str] = None
    root_cause: Optional[str] = None
    explanation: Optional[str] = None
    evidence: List[Evidence] = []


class AnomalySummary(BaseModel):
    total: int
    anomalies: int
    by_family: Dict[str, int]


class AnomalyDetectResponse(BaseResponse):
    product_id: str
    aps: str
    baseline_method: str
    effective_thresholds: AnomalyThresholds
    series: List[AnomalySeriesItem]
    summary: AnomalySummary
    warnings: List[str] = []


class AnomalyOptionsResponse(BaseResponse):
    products: List[str]
    aps_list: List[str]
    default_thresholds: AnomalyThresholds


class AnomalyMetricsResponse(BaseResponse):
    metrics: List[Dict[str, str]]
    warnings: List[str] = []


class AnomalyBatchResponse(BaseResponse):
    results: Optional[List[Dict[str, object]]] = None
    job_id: Optional[str] = None
    status: Optional[str] = None
    warnings: List[str] = []

File: webapp/frontend/src/api/types/anomaly.ts
import { BaseResponse } from "./common";

export interface Evidence {
  type: string;
  strength: number;
  detail: string;
  ref?: string | null;
}

export interface AnomalyThresholds {
  point: number;
  pattern: number;
  trend: number;
}

export interface AnomalyDetectRequest {
  product_id: string;
  aps: string;
  date_range?: [string, string] | null;
  thresholds?: AnomalyThresholds | null;
  threshold?: number | null;
  include_explanations?: boolean;
}

export interface AnomalySeriesItem {
  date: string;
  demand: number;
  expected: number;
  residual: number;
  anomaly_score: number;
  is_anomaly: boolean;
  anomaly_family?: string | null;
  anomaly_type?: string | null;
  root_cause?: string | null;
  explanation?: string | null;
  evidence: Evidence[];
}

export interface AnomalySummary {
  total: number;
  anomalies: number;
  by_family: Record<string, number>;
}

export interface AnomalyDetectResponse extends BaseResponse {
  product_id: string;
  aps: string;
  baseline_method: string;
  effective_thresholds: AnomalyThresholds;
  series: AnomalySeriesItem[];
  summary: AnomalySummary;
  warnings?: string[];
}

export interface AnomalyOptionsResponse extends BaseResponse {
  products: string[];
  aps_list: string[];
  default_thresholds: AnomalyThresholds;
}

export interface AnomalyBatchResponse extends BaseResponse {
  results?: AnomalyDetectResponse[];
  job_id?: string;
  status?: string;
  warnings?: string[];
}

export interface AnomalyMetricsResponse extends BaseResponse {
  metrics: { label: string; value: string }[];
  warnings?: string[];
}
